Bluetooth: has: Satisfy profile security requirement

MariuszSkamra opened the pull request:
This enables Secure Connections pairing (actually disables legacy)
to satisfy security requirements of the Hearing Access Profile.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -19,13 +19,45 @@
+	/* The security requirements for all characteristics is LE Security Mode 1 Level 2 */
+	BT_ASSERT(bt_conn_get_security(conn) > BT_SECURITY_L1);
+
+	if (bt_conn_enc_key_size(conn) < HAS_ENC_KEY_SIZE_MIN) {
+		return BT_GATT_ERR(BT_ATT_ERR_ENCRYPTION_KEY_SIZE);
+	}
```
Unfortunately since the security requirements for LE Audio does not match a security level, we can't just do `bt_conn_get_security` and do a check on the appropriate level. 

Due to this we'll need to implement a new function to return whether on not it is a secure connection. 
For Zephyr I don't think we need to worry about the BR/EDR. 

Not sure what do to with the `OOB method` verification.

MariuszSkamra replied:
I guess your comment address this line:
```
	/* The security requirements for all characteristics is LE Security Mode 1 Level 2 */
	BT_ASSERT(bt_conn_get_security(conn) > BT_SECURITY_L1);
```
This assert ensures the characteristic if BT_GATT_PERM_READ_ENCRYPT and/or BT_GATT_PERM_WRITE_ENCRYPT property bit is set and the characteristic is not accessed if there's no encryption.
The LE Security Mode 1 Level 2 requirement is defined in the specification so the match is valid even if more security requirements are defined.

Regarding `bt_conn_enc_key_size()` check is used to validate the encryption key size that shall be at least 128bit as defined is specification.

Thalley replied:
That is true, but I was referring to LE Audio basically using something like security level 2.5 (i.e. security level 2 + 128-bit key requirement and LE Secure Connections/OOB). 

Unless we can somehow move this permission check to GATT, how do you feel about introducing two new functions, 

`bt_conn_is_secure_connection(conn)` and `bt_conn_is_oob_secured(conn)` so that we can easily verify that it is either encrypted with LE Secure Connections, or by OOB? (`bt_conn_is_secure_connection` is probably poor choice, since it may give the wrong impression on people unfamiliar with LE secure connections).

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -19,13 +19,45 @@
+	 * • BR/EDR Secure Connections (if Cross-Transport Key Derivation is used)
+	 * • OOB method
+	 *
+	 * TODO: Check the encryption method.
+	 *       What error shall be returned if requirement above is not met??
+	 */
```
`BT_ATT_ERR_INSUFFICIENT_ENCRYPTION` perhaps?

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -19,13 +19,45 @@
+
 	BT_DBG("conn %p attr %p offset %d", (void *)conn, attr, offset);
 
+	err = check_security(conn);
```
Would it make more sense to add these additional security requirements to the GATT API, and have GATT handle it, like it currently handles the `BT_GATT_PERM_READ_` values?

jhedberg replied:
Is the only additional requirement a check for the encryption key size? I don't mind doing this on a GATT level (which seems better than what's right now proposed in this PR), but would it perhaps be even simpler to do this in Kconfig by adding something like `depends on BT_SMP_MIN_ENC_KEY_SIZE == 16`?

Thalley replied:
> Is the only additional requirement a check for the encryption key size? I don't mind doing this on a GATT level (which seems better than what's right now proposed in this PR), but would it perhaps be even simpler to do this in Kconfig by adding something like `depends on BT_SMP_MIN_ENC_KEY_SIZE == 16`?

From the PR description: 
> Access to all characteristics (...) shall require an encryption
> key with at least 128 bits of entropy, derived from any of the following:
> • LE Secure Connections
> • BR/EDR Secure Connections (if Cross-Transport Key Derivationd)
> • OOB method

So are 2 requirements: 128-bit key, and then one of the 3 methods used. 

We've used something like `depends on BT_SMP_MIN_ENC_KEY_SIZE == 16`  in the past in the form of `select BT_SMP_SC_PAIR_ONLY` (or `depends on BT_SMP_SC_PAIR_ONLY`) to satisfy the LE secure connection requirement. 

However, this has the downside of being too strict: Technically the requirement is only the service characteristic access, and using `BT_SMP_SC_PAIR_ONLY` or `BT_SMP_MIN_ENC_KEY_SIZE` is a device-wide requirement. In most cases it should be fine, but we are adding additional requirements for non-"LE Audio" services that way

For the key size, we could add an additional value for the GATT characeristics that contain the minimum required key length. For the other requirements (LE secure connection, BR/EDR secure connections or OOB) it is going to be harder.

jhedberg replied:
@Thalley is the method stuff really that complicated? The only remaining method is legacy LE pairing, so all you really need to be able to exclude is that (then anything else satisfies the requirement), right?

jhedberg replied:
... in that sense it's probably sufficient to add a boolean option (or flag) to the characteristic definition

jhedberg replied:
Actually, we already have that `BT_SMP_SC_PAIR_ONLY` (and I see you already referenced it). I don't think there's a problem with that being a device-wide requirement, since (unlike `BT_SMP_SC_ONLY`) you can still have characteristics which don't require encryption, mixed which characteristics which do. It's probably also better from a user experience perspective, since you don't get weird situations where you've been able to pair, but only part of the services work - I think it's better to just fail the pairing in that case.

jhedberg replied:
The same argument can also be used for the encryption key size, i.e. that it's less confusing for the user to fail pairing rather than have a mix of services that work and those that don't work. So in that sense this should be solvable with the existing Kconfig options.

Thalley replied:
I think that was also the conclusion last time we visited this issue, which is why we ended up doing `select BT_SMP_SC_PAIR_ONLY`. 

At the very least, it's is better to be more restrictive than not complying with security requirements :) 

`BT_SMP_SC_PAIR_ONLY` does not cover the OOB method though, and as mentioned I don't think it's worth dealing with the BR/EDR method. 

Personally I am fine with rejecting pairing that doesn't conform with the highest security requirement, even if it perhaps isn't considered the "GATT way". 

Any device that supports the LE Audio services will be advertising that, and in the case you are connecting to those devices, you should definitely be able to support LE Secure Connections and 128-bits anyhow.

MariuszSkamra replied:
So the conclusion is to enable `BT_SMP_SC_PAIR_ONLY` option.
The PR has been updated.

Thalley reviewed (commented):
This is something we discussed a long time ago in Zephyr, but didn't really solver properly.

MariuszSkamra committed:
Bluetooth: has: Satisfy profile security requirement

This enables Secure Connections pairing (actually disables legacy)
to satisfy security requirements of the Hearing Access Profile.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/Kconfig.has:
```diff
@@ -7,6 +7,7 @@
 menuconfig BT_HAS
 	bool "Hearing Access Service support [EXPERIMENTAL]"
 	select EXPERIMENTAL
+	select BT_SMP_SC_PAIR_ONLY
```
Should we also add `depends on BT_SMP_MIN_ENC_KEY_SIZE == 16`, or some other way of satisfying that?

MariuszSkamra replied:
Right!

MariuszSkamra replied:
@Thalley To be honest I was not aware of such Kconfig option

Thalley commented on subsys/bluetooth/audio/Kconfig.has:
```diff
@@ -7,6 +7,8 @@
 menuconfig BT_HAS
 	bool "Hearing Access Service support [EXPERIMENTAL]"
 	select EXPERIMENTAL
+	select BT_SMP_SC_PAIR_ONLY
+	depends on BT_SMP_MIN_ENC_KEY_SIZE = 16
```
The final question:

Should we use `select BT_SMP_SC_PAIR_ONLY` or `depends on BT_SMP_SC_PAIR_ONLY` where the former is going to be easier to use, but is inconsistent with `depends on BT_SMP_MIN_ENC_KEY_SIZE = 16`.


Previously we've in LE Audio have received significant pushback on using `depends on` when adding dependencies, because it hides Kconfig options in the `menuconfig` GUI, but it's hard to avoid for something like `depends on BT_SMP_MIN_ENC_KEY_SIZE = 16`. 

The alternative we were (forced) to adopt by a reviewer with regards to TX buffer count, were to add a 

`default 16 if BT_HAS` in `BT_SMP_MIN_ENC_KEY_SIZE` instead, and then add a `BUILD_ASSERT`, to actually ensure the value was set... 

Personally I like `depends on` a lot better, but I also do not use the GUI. 

@asbjornsabo @jhedberg @MariuszSkamra  What are you thoughts on this?

jhedberg replied:
`depends on` is usually the safer option. Btw, I recommend reading this section in the Zephyr documentation if you haven't already: https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#select-pitfalls

MariuszSkamra replied:
> The final question:
> 
> Should we use `select BT_SMP_SC_PAIR_ONLY` or `depends on BT_SMP_SC_PAIR_ONLY` where the former is going to be easier to use, but is inconsistent with `depends on BT_SMP_MIN_ENC_KEY_SIZE = 16`.

@Thalley I am not pedantic. Does it really matter if one dependency is defined with `select` and another `depends on` ?
Both are used together in many places across the Zephyr.

> `depends on` is usually the safer option. Btw, I recommend reading this section in the Zephyr documentation if you haven't already: https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#select-pitfalls

@jhedberg Thanks for the link, I'm aware that `select` does not resolve dependencies, but I saw it's already used in few places to enable `BT_SMP_SC_PAIR_ONLY`.

I don't have a opinion regarding this. If you say I should use `depends on` I'll change it.

Thalley replied:
Either way ensures that the Kconfig is enabled :)
